Add hetzner service discovery (currently only for Hetzner Cloud)

LKaemmerling opened the pull request:
This MR adds Hetzner (https://hetzner.com) service discovery to Prometheus. As written within the IRC this service discovery is written with extendibility in mind. The initial plan was to add only a service discovery for Hetzner Cloud instances (https://hetzner.cloud), but @brian-brazil found a more generic "Hetzner" way better as this allows to add Hetzner Dedicated Servers to the service discovery in the future too.

I plan to maintain this service discovery in the future and will also add support for Hetzner Dedicated servers in the future. 

cc @roidelapluie

brian-brazil commented on discovery/hetzner/hetzner.go:
```diff
@@ -0,0 +1,173 @@
+		return nil, err
+	}
+	opts := []hcloud.ClientOption{
+		hcloud.WithToken(string(conf.HTTPClientConfig.BearerToken)),
```
Leave this up to the round tripper - HTTPClientConfig is a black box to you, and this is also incorrect for the file.

brian-brazil commented on discovery/hetzner/hetzner.go:
```diff
@@ -0,0 +1,173 @@
+		hcloud.WithApplication("Prometheus", version.Version),
+		hcloud.WithHTTPClient(&http.Client{
+			Transport: rt,
+			Timeout:   5 * time.Duration(conf.RefreshInterval),
```
This seems a bit long

brian-brazil commented on discovery/hetzner/hetzner.go:
```diff
@@ -0,0 +1,173 @@
+
+	RefreshInterval model.Duration `yaml:"refresh_interval"`
+	Port            int            `yaml:"port"`
+	HcloudSDConfig  HcloudSDConfig `yaml:"hcloud"`
```
Why not put these values in directly?

LKaemmerling replied:
I wanted to separate them because they won't be valid for dedicated servers and I wanted to avoid having something like `hcloud_network` `robot_vswitch` and instead have a clean object structure.

brian-brazil replied:
Let's worry about that if/when it happens, and keep things simple for now. We generally don't nest for this sort of thing.

brian-brazil commented on discovery/hetzner/hetzner.go:
```diff
@@ -0,0 +1,173 @@
+			model.AddressLabel: model.LabelValue(net.JoinHostPort(server.PublicNet.IPv4.IP.String(), strconv.FormatUint(uint64(d.port), 10))),
+		}
+
+		if server.Image.Name == "" {
```
Don't munge, put out the raw metadata and leave it up to relabelling.

LKaemmerling replied:
The name is only set for images of type system so it could be empty when you have a snapshot or a backup in that case the description is set. Should i add a new label "hetzner_hcloud_image_description" ?

brian-brazil replied:
If it's empty, it's empty.

brian-brazil commented on discovery/hetzner/hetzner.go:
```diff
@@ -0,0 +1,173 @@
+	hcloudLabelDatacenter  = model.MetaLabelPrefix + "hcloud_datacenter"
+	hcloudLabelCPUCores    = model.MetaLabelPrefix + "hcloud_cpu_cores"
+	hcloudLabelCPUType     = model.MetaLabelPrefix + "hcloud_cpu_type"
+	hcloudLabelMemory      = model.MetaLabelPrefix + "hcloud_memory_size"
```
what's the unit?

brian-brazil commented on docs/configuration/configuration.md:
```diff
@@ -215,6 +215,11 @@ file_sd_configs:
   [ - <gce_sd_config> ... ]
 
+# List of Hetzner Cloud service discovery configurations.
+hcloud_sd_configs:
```
This needs updating.

brian-brazil commented on discovery/hetzner/hetzner.go:
```diff
@@ -0,0 +1,173 @@
+)
+
+const (
+	hcloudLabelID          = model.MetaLabelPrefix + "hcloud_id"
```
Metadata prefix should match the SD name

brian-brazil replied:
Id of what?

roidelapluie commented:
Can you clarify whether or not you plan to support something else than hcloud in the future?

LKaemmerling commented:
> Can you clarify whether or not you plan to support something else than hcloud in the future?

After talking with @brian-brazil in IRC this MR will include both Hetzner Cloud and Hetzner Robot support. I'm already working on it and just need to finish the Robot tests.

LKaemmerling commented:
@brian-brazil and @roidelapluie i just updated the MR to support both Hetzner Cloud Servers and Hetzner Robot (Dedicated) Servers. I used the openstack service discovery as a foundation.

roidelapluie commented on discovery/hetzner/hcloud.go:
```diff
@@ -0,0 +1,125 @@
+			Transport: rt,
+			Timeout:   time.Duration(conf.RefreshInterval),
+		}),
+		hcloud.WithPollInterval(time.Duration(conf.RefreshInterval)),
```
Why do we need this? Prometheus will do this.

roidelapluie commented on discovery/hetzner/hcloud.go:
```diff
@@ -0,0 +1,125 @@
+			Timeout:   time.Duration(conf.RefreshInterval),
+		}),
+		hcloud.WithPollInterval(time.Duration(conf.RefreshInterval)),
+		hcloud.WithEndpoint(conf.HCloudEndpoint),
```
should HCloudEndpoint be public? Why would users override it? We could make it private to limit it to tests.

roidelapluie commented on discovery/hetzner/hcloud.go:
```diff
@@ -0,0 +1,125 @@
+			hetznerLabelServerName:             model.LabelValue(server.Name),
+			hetznerLabelDatacenter:             model.LabelValue(server.Datacenter.Name),
+			hetznerLabelPublicIPv4:             model.LabelValue(server.PublicNet.IPv4.IP.String()),
+			hetznerLabelHCloudLocation:         model.LabelValue(server.Datacenter.Location.Name),
```
should we have the networkzone as a label too?

roidelapluie commented on discovery/hetzner/hcloud_test.go:
```diff
@@ -0,0 +1,99 @@
+	Mock *SDMock
+}
+
+func (s *HCloudSDTestSuite) TearDownSuite() {
+	s.Mock.ShutdownServer()
+}
+
```
```suggestion

```

roidelapluie commented on discovery/hetzner/mock_test.go:
```diff
@@ -0,0 +1,393 @@
+// Setup creates the mock server
+func (m *SDMock) Setup() {
+	m.Mux = http.NewServeMux()
+	m.Server = httptest.NewServer(m.Mux)
```
```suggestion
	m.Server = httptest.NewServer(m.Mux)
	m.t.Cleanup(m.Server.Close)
```

roidelapluie commented on discovery/hetzner/hcloud_test.go:
```diff
@@ -0,0 +1,99 @@
+func TestHCloudSDRefresh(t *testing.T) {
+	suite := &HCloudSDTestSuite{}
+	suite.SetupTest(t)
+	t.Cleanup(suite.TearDownSuite)
```
```suggestion
```

roidelapluie commented on discovery/hetzner/hetzner.go:
```diff
@@ -0,0 +1,142 @@
+	Port            int             `yaml:"port"`
+	Platform        HetznerPlatform `yaml:"platform"`
+	HCloudNetwork   string          `yaml:"hcloud_network"`
+	HCloudEndpoint  string          `yaml:"hcloud_endpoint"`
```
Why would users override that?

roidelapluie commented on discovery/hetzner/hetzner.go:
```diff
@@ -0,0 +1,142 @@
+}
+
+// SDConfig is the configuration for Hetzner based service discovery.
+type HcloudSDConfig struct {
+	Network  string `yaml:"network"`
+	Endpoint string `yaml:"endpoint"`
+}
```
```suggestion
```

roidelapluie commented on discovery/hetzner/hetzner.go:
```diff
@@ -0,0 +1,142 @@
+	), nil
+}
+
+func newRefresher(conf *SDConfig, l log.Logger) (refresher, error) {
```
is the unmarshalling tested?

roidelapluie commented on discovery/hetzner/robot.go:
```diff
@@ -0,0 +1,92 @@
+	"github.com/prometheus/prometheus/discovery/targetgroup"
+)
+
+// Discovery periodically performs Hetzner Cloud requests. It implements
+// the Discoverer interface.
+type RobotDiscovery struct {
```
```suggestion
// robotDiscovery periodically performs Hetzner Cloud requests. It implements
// the Discoverer interface.
type robotDiscovery struct {
```

roidelapluie commented on discovery/hetzner/robot.go:
```diff
@@ -0,0 +1,92 @@
+
+	for _, server := range servers {
+		labels := model.LabelSet{
+			hetznerLabelPlatform:   model.LabelValue(HetznerPlatformRobot),
```
```suggestion
			hetznerLabelPlatform:   model.LabelValue(hetznerPlatformRobot),
```

roidelapluie commented on discovery/hetzner/robot.go:
```diff
@@ -0,0 +1,92 @@
+
+	var servers serversList
+	err = json.NewDecoder(resp.Body).Decode(&servers)
+	var targets []model.LabelSet
```
```suggestion
	targets := make([]model.LabelSet, len(servers))
```

roidelapluie commented on discovery/hetzner/robot.go:
```diff
@@ -0,0 +1,92 @@
+	}
+
+	var servers serversList
+	err = json.NewDecoder(resp.Body).Decode(&servers)
```
you are not doing anything with the error

roidelapluie commented on discovery/hetzner/robot.go:
```diff
@@ -0,0 +1,92 @@
+	resp, err := d.client.Get(d.endpoint + "/server")
+	if err != nil {
+		return nil, err
+	}
```
defer resp.body close

roidelapluie commented on discovery/hetzner/robot.go:
```diff
@@ -0,0 +1,92 @@
+	return d, nil
+}
+func (d *RobotDiscovery) refresh(ctx context.Context) ([]*targetgroup.Group, error) {
+	resp, err := d.client.Get(d.endpoint + "/server")
```
no pagination?

LKaemmerling replied:
The Robot API has no pagination

roidelapluie commented on discovery/hetzner/robot_test.go:
```diff
@@ -0,0 +1,84 @@
+	Mock *SDMock
+}
+
+func (s *RobotSDTestSuite) TearDownSuite() {
+	s.Mock.ShutdownServer()
+}
+
```
```suggestion


```

roidelapluie commented on discovery/hetzner/robot_test.go:
```diff
@@ -0,0 +1,84 @@
+func TestRobotSDRefresh(t *testing.T) {
+	suite := &RobotSDTestSuite{}
+	suite.SetupTest(t)
+	t.Cleanup(suite.TearDownSuite)
```
```suggestion
```

roidelapluie commented on docs/configuration/configuration.md:
````diff
@@ -918,6 +922,77 @@ instance it is running on should have at least read-only permissions to the
+[ hcloud_network: <string> | default = "" ]
+
+```
+
````
You miss hetzner_sd_configs below in the alertmanager doc

roidelapluie commented on docs/configuration/configuration.md:
```diff
@@ -918,6 +922,77 @@ instance it is running on should have at least read-only permissions to the
+# The time after which the droplets are refreshed.
+[ refresh_interval: <duration> | default = 60s ]
+
+# The name or the ID of the network the servers are attached to
```
full stop

roidelapluie replied:
specify it is only for hcloud

roidelapluie commented on docs/configuration/configuration.md:
```diff
@@ -918,6 +922,77 @@ instance it is running on should have at least read-only permissions to the
+  [ <tls_config> ]
+
+# The Hetzner platform of entities that should be discovered.
+# One of robot or cloud.
```
```suggestion
# One of robot or hcloud.
```

what about we ue hcloud

roidelapluie commented on docs/configuration/configuration.md:
```diff
@@ -918,6 +922,77 @@ instance it is running on should have at least read-only permissions to the
+Additionally, if you configure the service discovery to look for a specific network
+the following meta labels are available:
+* `__meta_hetzner_hcloud_network`: the name of the network the server is attached to
+* `__meta_hetzner_hcloud_private_ipv4`: the private ipv4 of the server within this specific network
```
The EC2 datasource is using meta api to take the current network id. Should we do this here too @brian-brazil ?

brian-brazil replied:
Yeah, this sort of thing should be completely automatic.

LKaemmerling replied:
This will only work if the server where Prometheus is running will be in the same (private) network as the servers, otherwise our metadata service won’t recognize the network. A server within our cloud can be attached to 3 different networks, so you need to specify the network that you want to select.

roidelapluie replied:
thanks for the info

brian-brazil replied:
That sounds like business logic, so we should expose metadata rather than adding config options.

roidelapluie commented on docs/configuration/configuration.md:
```diff
@@ -918,6 +922,77 @@ instance it is running on should have at least read-only permissions to the
+* `__meta_hetzner_datacenter`: the name of the datacenter the server is located
+
+
+The labels below are only available for targets with `platform` set to `cloud`
```
```suggestion
The labels below are only available for targets with `platform` set to `cloud`:
```

roidelapluie commented on docs/configuration/configuration.md:
```diff
@@ -918,6 +922,77 @@ instance it is running on should have at least read-only permissions to the
+
+* `__meta_hetzner_server_id`: the id of the server
+* `__meta_hetzner_server_name`: the name of the server
+* `__meta_hetzner_ipv4`: the public ipv4 of the server
```
there is not ipv6?

roidelapluie commented on discovery/hetzner/hcloud_test.go:
```diff
@@ -0,0 +1,99 @@
+	"testing"
+)
+
+type HCloudSDTestSuite struct {
```
should that be public?

roidelapluie reviewed (commented):
Initial review, thanks

roidelapluie commented:
We miss the server status label (on/off)

roidelapluie commented:
We are missing robot labels:
```
  {                                                                                                                                        
    "server": {                                                                                                                            
      "server_ip": "1.1.1.1",                                                                                                       
      "server_number": 42,                                                                                                             
      "server_name": "a.b.c.d",                                                                                            
      "product": "PX60-SSD",                                                                                                               
      "dc": "FSN1-DC12",                                                                                                                   
      "traffic": "unlimited",                                                                                                              
      "flatrate": true,                                                                                                                    
      "status": "ready",                                                                                                                   
      "throttled": false,                                                                                                                  
      "cancelled": false,                                                                                                                  
      "paid_until": "2020-09-07",                                                                                                          
      "ip": [                                                                                                                              
        "1.1.1.1"                                                                                                                   
      ],                                                                                                                                   
      "subnet": [                
        {                         
          "ip": "2a01:XX::",
          "mask": "64"                                               
        }                                                            
  ```

LKaemmerling commented:
@roidelapluie and @brian-brazil I guess I have addressed everything with the latest commit. 
- Everything is no longer exported
- renamed the platform `cloud` to `hcloud`
- added additional robot fields (and added all fields to the documentation)
- added error handling for the robot API
- applied you suggestion regarding the mock server/testsuite
- `hcloudEndpoint` and `robotEndpoint` are no longer "user facing"
- avoided %v in fmt.Sprintf


The only "problematic" thing I see is with the comment regarding the `network configuration` from @brian-brazil (https://github.com/prometheus/prometheus/pull/7822#discussion_r472534105):
> That sounds like business logic, so we should expose metadata for relabelling rather than adding config options.

How could we expose the "networks" of a server if it has more than one network attached? On our cloud a server can have up to three networks attached ("Networks" is our product name for our private Network). The initial idea behind the configuration was that mostly the user is only interested in fetching/using one IP of the networks. Example: You have a Network `database-network (10.0.0.0/24)` for the database replication and another Network called `management-network (192.168.1.0/24)` for everything regarding managing/monitoring your servers. Prometheus now should only connect over the `management-network`. Therefore you specify `management-network` in your configuration and Prometheus always gets the correct private IP set within the label.

roidelapluie commented:
Create one  target per network/private IP and let the user drop the ones that are not interesting or create _network_ipv4_<network_name>: 1.1.1.1 label.

LKaemmerling commented:
> Create one target per network/private IP and let the user drop the ones that are not interesting or create _network_ipv4_<network_name>: 1.1.1.1 label.

I changed it, here is a sample screenshot with the generated labels:
![Bildschirmfoto 2020-08-19 um 09 25 46](https://user-images.githubusercontent.com/4281581/90605455-a8e8e700-e1fe-11ea-964c-13d27781d2c9.png)

roidelapluie commented on discovery/config/config.go:
```diff
@@ -64,6 +65,8 @@ type ServiceDiscoveryConfig struct {
 	// List of Triton service discovery configurations.
 	TritonSDConfigs []*triton.SDConfig `yaml:"triton_sd_configs,omitempty"`
+	// List of Hetzner service discovery configurations.
+	HetznerSDConfigs []*hetzner.SDConfig `yaml:"hetzner_sd_configs,omitempty"`
```
can you move it  above ?

roidelapluie commented:
Can you adapt resolveFilepaths in config? and add the discovery to config/testdata/conf.good.yml

roidelapluie commented on discovery/hetzner/hcloud.go:
```diff
@@ -0,0 +1,122 @@
+	if err != nil {
+		return nil, err
+	}
+	networks, err := d.client.Network.All(ctx)
```
The API seems self contained, is this needed?

LKaemmerling replied:
The servers endpoint just answers with the network ID (https://docs.hetzner.cloud/#servers-get-a-server) so we need to get the list of all networks first

roidelapluie commented on discovery/hetzner/hcloud.go:
```diff
@@ -0,0 +1,122 @@
+			hetznerLabelServerName:   model.LabelValue(server.Name),
+			hetznerLabelDatacenter:   model.LabelValue(server.Datacenter.Name),
+			hetznerLabelPublicIPv4:   model.LabelValue(server.PublicNet.IPv4.IP.String()),
+			hetznerLabelPublicIPv6:   model.LabelValue(server.PublicNet.IPv6.Network.String()),
```
Do you mean network or IP?

LKaemmerling replied:
Every server (dedicated and cloud) has a /64 IPv6 Net and we never know which IP within this network is correctly configured

roidelapluie replied:
if the api returns both let's have both net and ip as label

LKaemmerling replied:
The API does not return an IP, it always return the IPv6 Net. I will rename the Label to "public_ipv6_net"

roidelapluie replied:
https://godoc.org/github.com/hetznercloud/hcloud-go/hcloud#ServerPublicNetIPv6 has ip which confused me

LKaemmerling replied:
Yeah on cloud servers we assume (or at least within the base images it is) that the first IP within the net is configured everytime.

roidelapluie commented on discovery/hetzner/hcloud_test.go:
```diff
@@ -0,0 +1,102 @@
+			"__meta_hetzner_server_id":                               model.LabelValue("42"),
+			"__meta_hetzner_server_name":                             model.LabelValue("my-server"),
+			"__meta_hetzner_server_status":                           model.LabelValue("running"),
+			"__meta_hetzner_ipv4":                                    model.LabelValue("1.2.3.4"),
```
```suggestion
			"__meta_hetzner_public_ipv4":                                    model.LabelValue("1.2.3.4"),
```

roidelapluie commented on discovery/hetzner/hcloud_test.go:
```diff
@@ -0,0 +1,102 @@
+			"__meta_hetzner_server_name":                             model.LabelValue("my-server"),
+			"__meta_hetzner_server_status":                           model.LabelValue("running"),
+			"__meta_hetzner_ipv4":                                    model.LabelValue("1.2.3.4"),
+			"__meta_hetzner_ipv6":                                    model.LabelValue("2001:db8::/64"),
```
```suggestion
			"__meta_hetzner_public_ipv6":                                    model.LabelValue("2001:db8::/64"),
```

roidelapluie commented on discovery/hetzner/hcloud_test.go:
```diff
@@ -0,0 +1,102 @@
+			"__meta_hetzner_datacenter":                              model.LabelValue("fsn1-dc8"),
+			"__meta_hetzner_hcloud_image_name":                       model.LabelValue("ubuntu-20.04"),
+			"__meta_hetzner_hcloud_image_description":                model.LabelValue("Ubuntu 20.04 Standard 64 bit"),
+			"__meta_hetzner_hcloud_datacenter_location":              model.LabelValue("fsn1"),
```
```suggestion
			"__meta_hetzner_datacenter_location":              model.LabelValue("fsn1"),
```

to match __meta_hetzner_datacenter

LKaemmerling replied:
The Location is only returned from the Cloud API. The Robot API doesn't know about "locations" at all. So I moved it to a cloud-specific label.

roidelapluie replied:
ack

roidelapluie commented:
For ease of use we usually provide a port: in our SD configs.

LKaemmerling commented:
> For ease of use we usually provide a port: in our SD configs.

If i see this correctly, this is already implemented

roidelapluie commented on discovery/hetzner/robot.go:
```diff
@@ -0,0 +1,117 @@
+		}
+		if len(server.Server.Subnet) > 0 {
+			for _, subnet := range server.Server.Subnet {
+				if subnet.Mask == "64" {
```
why is that needed? Is there IPv4 subnets too?

LKaemmerling replied:
You might get IPv4 subnets through the endpoint when a server has multiple subnets. I would like to keep them out of the labels in the first version of the service discovery.

brian-brazil replied:
This seems a tad business logicy - what if you give our different mask sizes in future?

roidelapluie commented on discovery/hetzner/robot_test.go:
```diff
@@ -0,0 +1,84 @@
+			"__meta_hetzner_server_name":     model.LabelValue("server1"),
+			"__meta_hetzner_server_status":   model.LabelValue("ready"),
+			"__meta_hetzner_public_ipv4":     model.LabelValue("123.123.123.123"),
+			"__meta_hetzner_public_ipv6_net": model.LabelValue("2a01:4f8:111:4221::/64"),
```
```suggestion
			"__meta_hetzner_public_ipv6_network": model.LabelValue("2a01:4f8:111:4221::/64"),
```

roidelapluie commented on docs/configuration/configuration.md:
```diff
@@ -918,6 +922,82 @@ instance it is running on should have at least read-only permissions to the
+* `__meta_hetzner_hcloud_disk_size_gb`: size of disk the server has (in GB)
+
+Additionally, if you configure the service discovery to look for a specific network
+the following meta labels are available:
```
is this still true? or do we provide everything?

roidelapluie commented on discovery/hetzner/robot_test.go:
```diff
@@ -0,0 +1,84 @@
+			"__meta_hetzner_server_status": model.LabelValue("in process"),
+			"__meta_hetzner_public_ipv4":   model.LabelValue("123.123.123.124"),
+			"__meta_hetzner_datacenter":    model.LabelValue("fsn1-dc10"),
+			"__meta_hetzner_robot_product": model.LabelValue("X5"),
```
You still miss fields:
flatrate, canceled, traffic..

LKaemmerling replied:
Does that fields really brings that to the user?  Maybe this is something that could be added later on.

roidelapluie replied:
Usually SD is just a pass-through. Canceled might bring added value, I don't know what is it.

roidelapluie replied:
But it sound like I would like to drop the targets that are marked as canceled ;)

LKaemmerling replied:
I just added a label for this.

roidelapluie commented on discovery/hetzner/hetzner.go:
```diff
@@ -0,0 +1,153 @@
+var DefaultSDConfig = SDConfig{
+	Port:            80,
+	RefreshInterval: model.Duration(60 * time.Second),
+	hcloudEndpoint:  "",
```
this is not needed

roidelapluie commented on discovery/hetzner/hetzner.go:
```diff
@@ -0,0 +1,153 @@
+	RefreshInterval model.Duration `yaml:"refresh_interval"`
+	Port            int            `yaml:"port"`
+	Platform        platform       `yaml:"platform"`
+	hcloudEndpoint  string
```
can you comment that this is for testing only?

roidelapluie commented on discovery/hetzner/hetzner.go:
```diff
@@ -0,0 +1,153 @@
+	case hetznerPlatformRobot, hetznerPlatformHcloud:
+		return nil
+	default:
+		return errors.Errorf("unknown Hetzner SD platform %q", *c)
```
```suggestion
		return errors.Errorf("unknown platform %q", *c)
```

roidelapluie commented on discovery/hetzner/hetzner.go:
```diff
@@ -0,0 +1,153 @@
+
+// NewDiscovery returns a new Discovery which periodically refreshes its targets.
+func NewDiscovery(conf *SDConfig, logger log.Logger) (*refresh.Discovery, error) {
+
```
extra white line

roidelapluie commented on docs/configuration/configuration.md:
```diff
@@ -918,6 +922,80 @@ instance it is running on should have at least read-only permissions to the
+* `__meta_hetzner_hcloud_datacenter_location`: the name of the location the server is located
+* `__meta_hetzner_hcloud_datacenter_location_network_zone`: the name of the network zone the server is located
+* `__meta_hetzner_hcloud_server_type`: the name of the type of the server
+* `__meta_hetzner_hcloud_cpu_cores`: count of CPU cores the server has
```
```suggestion
* `__meta_hetzner_hcloud_cpu_cores`: the count of CPU cores the server has
```

roidelapluie commented on docs/configuration/configuration.md:
```diff
@@ -918,6 +922,80 @@ instance it is running on should have at least read-only permissions to the
+* `__meta_hetzner_hcloud_server_type`: the name of the type of the server
+* `__meta_hetzner_hcloud_cpu_cores`: count of CPU cores the server has
+* `__meta_hetzner_hcloud_cpu_type`: the type of the CPU (shared or dedicated)
+* `__meta_hetzner_hcloud_memory_size_gb`: amount of memory the server has (in GB)
```
the*

roidelapluie commented on docs/configuration/configuration.md:
```diff
@@ -918,6 +922,80 @@ instance it is running on should have at least read-only permissions to the
+* `__meta_hetzner_hcloud_cpu_cores`: count of CPU cores the server has
+* `__meta_hetzner_hcloud_cpu_type`: the type of the CPU (shared or dedicated)
+* `__meta_hetzner_hcloud_memory_size_gb`: amount of memory the server has (in GB)
+* `__meta_hetzner_hcloud_disk_size_gb`: size of disk the server has (in GB)
```
the*

roidelapluie commented on docs/configuration/configuration.md:
```diff
@@ -918,6 +922,80 @@ instance it is running on should have at least read-only permissions to the
+
+# The Hetzner platform of entities that should be discovered.
+# One of robot or hcloud.
+platform: <string>
```
I'd move it to the top

roidelapluie commented on docs/configuration/configuration.md:
```diff
@@ -918,6 +922,80 @@ instance it is running on should have at least read-only permissions to the
+
+# The time after which the droplets are refreshed.
+[ refresh_interval: <duration> | default = 60s ]
+
```
extra blank line

roidelapluie reviewed (commented):
Next round, getting close ;)

roidelapluie commented on discovery/hetzner/robot.go:
```diff
@@ -0,0 +1,120 @@
+func (d *robotDiscovery) refresh(ctx context.Context) ([]*targetgroup.Group, error) {
+	resp, err := d.client.Get(d.endpoint + "/server")
+	if err != nil {
+		defer resp.Body.Close()
```
the defer should happend when wee is not nil also please call io.Copy(ioutil.Discard, resp.Body)

roidelapluie commented on discovery/hetzner/hcloud.go:
```diff
@@ -0,0 +1,122 @@
+		for _, privateNet := range server.PrivateNet {
+			for _, network := range networks {
+				if privateNet.Network.ID == network.ID {
+					networkLabel := fmt.Sprintf("%s_%s", hetznerLabelHcloudPrivateIPv4, strings.ReplaceAll(network.Name, "-", "_"))
```
```suggestion
					networkLabel := := hetznerLabelHcloudPrivateIPv4 + strutil.SanitizeLabelName(network.Name)
```

roidelapluie commented on docs/configuration/configuration.md:
```diff
@@ -918,6 +922,79 @@ instance it is running on should have at least read-only permissions to the
+
+Hetzner SD configurations allow retrieving scrape targets from [Hetzner Cloud](https://www.hetzner.cloud/)
+Server API.
+This service discovery uses the public IPv4 address by default, by that can be
```
```suggestion
This service discovery uses the public IPv4 address by default, but that can be
```

roidelapluie commented on docs/configuration/configuration.md:
```diff
@@ -918,6 +922,79 @@ instance it is running on should have at least read-only permissions to the
+Hetzner SD configurations allow retrieving scrape targets from [Hetzner Cloud](https://www.hetzner.cloud/)
+Server API.
+This service discovery uses the public IPv4 address by default, by that can be
+changed with relabelling, as demonstrated in [the Prometheus hetzner-sd
```
relabeling

roidelapluie commented on docs/configuration/configuration.md:
```diff
@@ -918,6 +922,79 @@ instance it is running on should have at least read-only permissions to the
+configuration file](/documentation/examples/prometheus-hetzner.yml).
+
+The following meta labels are available on all targets during [relabeling](#relabel_config):
+
```
extra line

roidelapluie commented on discovery/hetzner/robot.go:
```diff
@@ -0,0 +1,123 @@
+}
+func (d *robotDiscovery) refresh(ctx context.Context) ([]*targetgroup.Group, error) {
+	resp, err := d.client.Get(d.endpoint + "/server")
+	defer resp.Body.Close()
+	if err != nil {
+		io.Copy(ioutil.Discard, resp.Body)
+		return nil, err
+	}
```
```suggestion
	if err != nil {
		return nil, err
	}
	defer func(){
		io.Copy(ioutil.Discard, resp.Body)
		resp.Body.Close()
	}
```

roidelapluie commented on discovery/hetzner/robot.go:
```diff
@@ -0,0 +1,124 @@
+func (d *robotDiscovery) refresh(ctx context.Context) ([]*targetgroup.Group, error) {
+	resp, err := d.client.Get(d.endpoint + "/server")
+	if err != nil {
+		io.Copy(ioutil.Discard, resp.Body)
```
this is not needed, if err is nil tis would panic

roidelapluie commented on discovery/hetzner/robot.go:
```diff
@@ -0,0 +1,124 @@
+	resp, err := d.client.Get(d.endpoint + "/server")
+	if err != nil {
+		io.Copy(ioutil.Discard, resp.Body)
+		defer resp.Body.Close()
```
this would panic

roidelapluie commented on discovery/hetzner/robot.go:
```diff
@@ -0,0 +1,124 @@
+		defer resp.Body.Close()
+		return nil, err
+	}
+	defer resp.Body.Close()
```
you do not defer the discard

brian-brazil commented on docs/configuration/configuration.md:
```diff
@@ -918,6 +922,77 @@ instance it is running on should have at least read-only permissions to the
+# mutually exclusive.
+# password and password_file are mutually exclusive.
+
+# Optional HTTP basic authentication information, required when platform is robot
```
Don't forget your full stops.

brian-brazil commented on discovery/hetzner/hetzner.go:
```diff
@@ -0,0 +1,150 @@
+	RefreshInterval model.Duration `yaml:"refresh_interval"`
+	Port            int            `yaml:"port"`
+	Platform        platform       `yaml:"platform"`
+	hcloudEndpoint  string         // for tests only
```
Comments should be full sentences, so capital letters and full stops.

brian-brazil commented on discovery/hetzner/robot.go:
```diff
@@ -0,0 +1,124 @@
+		return nil, err
+	}
+	defer func() {
+		io.Copy(ioutil.Discard, resp.Body)
```
Why do you need this? The decoder should consume everything.

roidelapluie replied:
We do this everywhere in the discover code base.

roidelapluie replied:
Exhaust every request body before closing it (#5166) (#5479)

    From the documentation:
    > The default HTTP client's Transport may not
    > reuse HTTP/1.x "keep-alive" TCP connections if the Body is
    > not read to completion and closed.

    This effectively enable keep-alive for the fixed requests.

abursavich replied:
@bradfitz doesn't mention the need to consume the entire body here in his boilerplate for the right way to make an HTTP request: https://github.com/bradfitz/exp-httpclient/blob/master/problems.md

If I'm reading the http code correctly, which I might not be... Ignoring cases where responses don't have bodies or when there's a protocol switch... It looks like the `readTransfer` [wraps an underlying reader](https://github.com/golang/go/blob/go1.15/src/net/http/transfer.go#L545-L565) with a `body` whose [Close method](https://github.com/golang/go/blob/go1.15/src/net/http/transfer.go#L961-L1000) reads until EOF (problem solved, right?)... But then `persistConn` [wraps that body in a `bodyEOFSignal`](https://github.com/golang/go/blob/go1.15/src/net/http/transport.go#L2121-L2141), whose job it is to keep track of reading until EOF but is incapable of observing that the `body` underneath is reading to EOF on Close (in the happy path)... So I think the http package is doing all the work of reading to EOF and blocking until it's done on `resp.Body.Close()`, but then it's closing the connection anyway.

Long story short, I think you need to manually read until EOF if you want to reuse a persistent connection in the event of a decoding failure (or extra whitespace at after the JSON object).

brian-brazil commented on discovery/hetzner/robot.go:
```diff
@@ -0,0 +1,124 @@
+
+			model.AddressLabel: model.LabelValue(net.JoinHostPort(server.Server.ServerIP, strconv.FormatUint(uint64(d.port), 10))),
+		}
+		if len(server.Server.Subnet) > 0 {
```
This if is redundant, the for will cover it.

LKaemmerling commented:
@brian-brazil I changed all your comments excerpt this https://github.com/prometheus/prometheus/pull/7822#discussion_r474128596 and I tried to find a future proof solution for 
https://github.com/prometheus/prometheus/pull/7822#discussion_r474209172. Hopefully, everything is fine now :)
